// mdb-cli/receivers/backend-receiver.js
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from) {
    for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
        to[j] = from[i];
    return to;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var open_1 = __importDefault(require("open"));
var inquirer_1 = require("inquirer");
var config_1 = __importDefault(require("../config"));
var receiver_1 = __importDefault(require("./receiver"));
var output_color_1 = require("../models/output-color");
var project_status_1 = require("../models/project-status");
var ftp_publish_strategy_1 = __importDefault(require("./strategies/publish/ftp-publish-strategy"));
var http_wrapper_1 = __importDefault(require("../utils/http-wrapper"));
var helpers_1 = __importDefault(require("../helpers"));
var cli_status_1 = require("../models/cli-status");
var BackendReceiver = /** @class */ (function (_super) {
    __extends(BackendReceiver, _super);
    function BackendReceiver(context) {
        var _this = _super.call(this, context) || this;
        _this.starterCode = '';
        _this.projectName = '';
        _this._publishRetries = 0;
        _this.context.authenticateUser();
        _this.options = {
            hostname: config_1.default.host,
            headers: { Authorization: "Bearer " + _this.context.userToken }
        };
        _this.context.registerNonArgFlags(['follow', 'ftp', 'ftp-only', 'open', 'remove', 'test', 'help', 'override']);
        _this.context.registerFlagExpansions({
            '-f': '--follow',
            '-n': '--name',
            '-o': '--open',
            '-p': '--platform',
            '-rm': '--remove',
            '-t': '--test',
            '-h': '--help'
        });
        _this.flags = _this.context.getParsedFlags();
        _this.args = _this.context.args;
        return _this;
    }
    BackendReceiver.prototype.list = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.result.liveTextLine('Fetching backend projects...');
                        return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _a.sent();
                        if (projects.length) {
                            result = projects.map(function (p) {
                                var technologyMeta = p.projectMeta.find(function (m) { return m.metaKey === '_backend_technology'; });
                                var technology = technologyMeta ? technologyMeta.metaValue : undefined;
                                var portMeta = p.projectMeta.find(function (m) { return m.metaKey === '_container_port'; });
                                var port = portMeta ? portMeta.metaValue : undefined;
                                var isUp = !!port;
                                var url = !!p.domainName ? "http://" + p.domainName : config_1.default.projectsDomain + ":" + port;
                                var deletedFromFTP = p.projectMeta.some(function (m) { return m.metaKey === '_uploaded_to_ftp' && m.metaValue === '0'; });
                                return {
                                    'Project Name': p.projectName,
                                    'Published': new Date(p.publishDate).toLocaleString(),
                                    'Edited': new Date(p.editDate).toLocaleString(),
                                    'Technology': technology,
                                    'Repository': p.repoUrl ? p.repoUrl : '-',
                                    'URL': isUp && !deletedFromFTP ? url : 'Unavailable'
                                };
                            });
                            this.result.addTable(result);
                        }
                        else {
                            this.result.addTextLine('You don\'t have any projects yet.');
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.getBackendProjects = function () {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.options.path = '/project';
                        return [4 /*yield*/, this.http.get(this.options)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, JSON.parse(result.body)
                                .filter(function (p) { return p.status === project_status_1.ProjectStatus.BACKEND; })
                                .sort(function (a, b) { return a.editDate < b.editDate; })];
                }
            });
        });
    };
    BackendReceiver.prototype.init = function (starterCode) {
        return __awaiter(this, void 0, void 0, function () {
            var initInCurrentFolder, projectPath, options, choices, err_1, initResult;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        initInCurrentFolder = this.context.args.some(function (arg) { return arg === '.'; });
                        if (initInCurrentFolder && fs_1.default.readdirSync(process.cwd()).length !== 0) {
                            return [2 /*return*/, this.result.addAlert(output_color_1.OutputColor.Red, 'Error', 'Destination path `.` already exists and is not an empty directory.')];
                        }
                        projectPath = process.cwd();
                        return [4 /*yield*/, this._getBackendStartersOptions()];
                    case 1:
                        options = _a.sent();
                        choices = this._buildBackendStartersList(options);
                        if (!starterCode) return [3 /*break*/, 2];
                        this.starterCode = starterCode;
                        this.projectName = this.flags.name || this.starterCode;
                        return [3 /*break*/, 4];
                    case 2: return [4 /*yield*/, this.chooseStarter(choices, options)];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4:
                        if (!!initInCurrentFolder) return [3 /*break*/, 9];
                        return [4 /*yield*/, this.checkProjectNameExists()];
                    case 5:
                        _a.sent();
                        projectPath = path_1.default.join(process.cwd(), this.projectName);
                        _a.label = 6;
                    case 6:
                        _a.trys.push([6, 8, , 9]);
                        return [4 /*yield*/, helpers_1.default.eraseDirectories(projectPath)];
                    case 7:
                        _a.sent();
                        return [3 /*break*/, 9];
                    case 8:
                        err_1 = _a.sent();
                        return [2 /*return*/, this.result.addAlert(output_color_1.OutputColor.Red, 'Error', err_1)];
                    case 9:
                        this.result.addTextLine("Project starter will be downloaded to " + projectPath + " folder");
                        return [4 /*yield*/, this.downloadProjectStarter(projectPath)];
                    case 10:
                        initResult = _a.sent();
                        this.result.addAlert(output_color_1.OutputColor.Green, 'Success', initResult);
                        this.context.mdbConfig.setValue('projectName', this.projectName);
                        this.context.mdbConfig.setValue('meta.starter', this.starterCode);
                        this.context.mdbConfig.setValue('meta.type', 'backend');
                        this.context.mdbConfig.setValue('hash', helpers_1.default.generateRandomString());
                        this.context.mdbConfig.save(projectPath);
                        this.context._loadPackageJsonConfig(projectPath);
                        if (!(this.context.packageJsonConfig.scripts && this.context.packageJsonConfig.scripts.test)) return [3 /*break*/, 12];
                        return [4 /*yield*/, helpers_1.default.createJenkinsfile(projectPath, true)];
                    case 11:
                        _a.sent();
                        _a.label = 12;
                    case 12: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.chooseStarter = function (choices, options) {
        return __awaiter(this, void 0, void 0, function () {
            var promptShownCount, _a, project;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        promptShownCount = 0;
                        _b.label = 1;
                    case 1:
                        if (promptShownCount++ >= 10) {
                            return [2 /*return*/, this.result.addTextLine('Please run `mdb starter ls` to see available packages.')];
                        }
                        _a = this;
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project to initialize', choices)];
                    case 2:
                        _a.starterCode = _b.sent();
                        this.projectName = this.flags.name || this.starterCode;
                        project = options.find(function (o) { return o.code === _this.starterCode; });
                        if (project.available)
                            return [3 /*break*/, 4];
                        else
                            this.result.liveAlert(output_color_1.OutputColor.Yellow, 'Warning!', "You cannot create this project. Please visit https://mdbootstrap.com/my-orders/ and make sure it is available for you.");
                        _b.label = 3;
                    case 3:
                        if (promptShownCount <= 10) return [3 /*break*/, 1];
                        _b.label = 4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype._getBackendStartersOptions = function () {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.options.path = "/packages/starters?type=backend" + (!this.flags.all ? '&available=true' : '');
                        return [4 /*yield*/, this.http.get(this.options)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, JSON.parse(result.body)];
                }
            });
        });
    };
    BackendReceiver.prototype._buildBackendStartersList = function (options) {
        var starters = options.reduce(function (res, curr) {
            res[curr.category + " " + curr.license] = res[curr.category + " " + curr.license] || [];
            res[curr.category + " " + curr.license].push({
                name: curr.displayName,
                short: curr.code,
                value: curr.code
            });
            return res;
        }, {});
        return Object.keys(starters).reduce(function (res, curr) {
            // @ts-ignore
            res.push.apply(res, __spreadArray([new inquirer_1.Separator("---- " + curr + " ----")], starters[curr]));
            return res;
        }, []);
    };
    BackendReceiver.prototype.checkProjectNameExists = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projectPath, confirmed, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        projectPath = path_1.default.join(process.cwd(), this.projectName);
                        if (!fs_1.default.existsSync(projectPath)) return [3 /*break*/, 4];
                        return [4 /*yield*/, helpers_1.default.createConfirmationPrompt("Folder " + this.projectName + " already exists, do you want to rename project you are creating now?", true)];
                    case 1:
                        confirmed = _b.sent();
                        if (!confirmed) return [3 /*break*/, 4];
                        _a = this;
                        return [4 /*yield*/, helpers_1.default.createTextPrompt('Enter new project name', 'Project name must not be empty.')];
                    case 2:
                        _a.projectName = _b.sent();
                        return [4 /*yield*/, this.checkProjectNameExists()];
                    case 3:
                        _b.sent();
                        _b.label = 4;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.downloadProjectStarter = function (projectPath) {
        return __awaiter(this, void 0, void 0, function () {
            var result;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.options.path = "/packages/download/" + this.starterCode;
                        return [4 /*yield*/, helpers_1.default.downloadFromFTP(this.http, this.options, projectPath)];
                    case 1:
                        result = _a.sent();
                        return [2 /*return*/, result];
                }
            });
        });
    };
    BackendReceiver.prototype.publish = function () {
        return __awaiter(this, void 0, void 0, function () {
            var platform, isNodePlatform, supportedPlatforms, packageJsonEmpty, result, e_1, packageJsonEmpty_1, result, e_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.flags.platform) {
                            this.context.mdbConfig.setValue('backend.platform', this.flags.platform);
                            this.context.mdbConfig.setValue('meta.type', 'backend');
                            this.context.mdbConfig.save();
                        }
                        if (!this.context.mdbConfig.getValue('hash')) {
                            this.context.mdbConfig.setValue('hash', helpers_1.default.generateRandomString());
                            this.context.mdbConfig.save();
                        }
                        platform = this.flags.platform || this.context.mdbConfig.getValue('backend.platform');
                        if (!platform) {
                            throw new Error('--platform flag is required when publishing backend projects!');
                        }
                        isNodePlatform = platform.includes('node');
                        if (isNodePlatform) {
                            this.result.liveAlert(output_color_1.OutputColor.Blue, 'Info', 'In order for your app to run properly you need to configure it so that it listens on port 3000. It is required for internal port mapping. The URL that your app is available at, will be provided to you after successful publish.');
                        }
                        supportedPlatforms = config_1.default.backendTechnologies;
                        if (!supportedPlatforms.includes(platform)) {
                            throw new Error("This technology is not supported. Allowed technologies: " + supportedPlatforms.join(', '));
                        }
                        packageJsonEmpty = this.context.packageJsonConfig.name === undefined;
                        if (!(packageJsonEmpty && isNodePlatform)) return [3 /*break*/, 5];
                        this.result.liveTextLine('package.json file is required. Creating...');
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.createPackageJson()];
                    case 2:
                        result = _a.sent();
                        this.context._loadPackageJsonConfig();
                        this.result.liveTextLine(result);
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _a.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', e_1);
                        return [2 /*return*/];
                    case 4:
                        packageJsonEmpty_1 = this.context.packageJsonConfig.name === undefined;
                        if (packageJsonEmpty_1) {
                            throw new Error('package.json file is required.'); // in case Ctrl+C
                        }
                        _a.label = 5;
                    case 5:
                        if (!this.flags.test) return [3 /*break*/, 9];
                        _a.label = 6;
                    case 6:
                        _a.trys.push([6, 8, , 9]);
                        return [4 /*yield*/, this.runTests()];
                    case 7:
                        result = _a.sent();
                        this.result.addAlert(output_color_1.OutputColor.Green, 'Success', result);
                        return [3 /*break*/, 9];
                    case 8:
                        e_2 = _a.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', e_2);
                        return [2 /*return*/];
                    case 9:
                        if (!!this.getProjectName()) return [3 /*break*/, 11];
                        return [4 /*yield*/, this.askForProjectName()];
                    case 10:
                        _a.sent();
                        _a.label = 11;
                    case 11: return [4 /*yield*/, this._handlePublication(packageJsonEmpty, isNodePlatform)];
                    case 12:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.createPackageJson = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.context.loadPackageManager()];
                    case 1:
                        _a.sent();
                        if (this.context.packageManager)
                            return [2 /*return*/, this.context.packageManager.init(process.cwd())];
                        return [2 /*return*/, Promise.reject('Failed to load the package manager.')];
                }
            });
        });
    };
    BackendReceiver.prototype.runTests = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.context.loadPackageManager()];
                    case 1:
                        _a.sent();
                        if (this.context.packageManager)
                            return [2 /*return*/, this.context.packageManager.test()];
                        return [2 /*return*/, Promise.reject('Failed to load the package manager.')];
                }
            });
        });
    };
    BackendReceiver.prototype._handlePublication = function (packageJsonEmpty, isNodePlatform) {
        return __awaiter(this, void 0, void 0, function () {
            var strategy, response, _a, message, url, e_3, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        this._publishRetries++;
                        if (this._publishRetries > 5) {
                            return [2 /*return*/, this.result.addAlert(output_color_1.OutputColor.Red, 'Error', 'Too many retries. Try again running the publish command.')];
                        }
                        strategy = new ftp_publish_strategy_1.default(this.context, this.result);
                        _c.label = 1;
                    case 1:
                        _c.trys.push([1, 3, , 8]);
                        return [4 /*yield*/, strategy.publish()];
                    case 2:
                        response = _c.sent();
                        _a = JSON.parse(response.body), message = _a.message, url = _a.url;
                        this.result.addTextLine(message);
                        if (this.flags.open && !!url)
                            open_1.default(url);
                        return [3 /*break*/, 8];
                    case 3:
                        e_3 = _c.sent();
                        if (!(e_3.statusCode === cli_status_1.CliStatus.CONFLICT && e_3.message.includes('project name'))) return [3 /*break*/, 6];
                        this.result.liveAlert(output_color_1.OutputColor.Red, 'Error', e_3.message);
                        _b = this;
                        return [4 /*yield*/, helpers_1.default.createTextPrompt('Enter new project name', 'Project name must not be empty.')];
                    case 4:
                        _b.projectName = _c.sent();
                        if (!packageJsonEmpty && isNodePlatform) {
                            this.context.setPackageJsonValue('name', this.projectName);
                        }
                        this.context.mdbConfig.setValue('projectName', this.projectName);
                        this.context.mdbConfig.save();
                        return [4 /*yield*/, this._handlePublication(packageJsonEmpty, isNodePlatform)];
                    case 5:
                        _c.sent();
                        return [3 /*break*/, 7];
                    case 6:
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not publish: " + (e_3.message || e_3));
                        _c.label = 7;
                    case 7: return [2 /*return*/];
                    case 8:
                        this.result.addAlert(output_color_1.OutputColor.Blue, 'Info', 'Since we need to install dependencies and run your app, it may take a few moments until it will be available.');
                        this.result.addTextLine('');
                        this.result.addAlert(output_color_1.OutputColor.Blue, 'Info', 'Your URL has been generated based on your username and project name. You can change it by providing the (sub)domain of your choice by running the following command: `mdb config domain <name>`.');
                        return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.delete = function (projectToDelete) {
        if (projectToDelete === void 0) { projectToDelete = this.flags.name; }
        return __awaiter(this, void 0, void 0, function () {
            var projects, choices, projectName, _a, projectExists, name, _b, query, result, err_2;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _c.sent();
                        if (projects.length === 0) {
                            this.result.addTextLine('You don\'t have any projects yet.');
                            return [2 /*return*/, false];
                        }
                        choices = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = projectToDelete;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', choices)];
                    case 2:
                        _a = (_c.sent());
                        _c.label = 3;
                    case 3:
                        projectName = _a;
                        projectExists = projects.some(function (p) { return p.projectName === projectName; });
                        if (!projectExists) {
                            this.result.addTextLine("Project " + projectName + " not found.");
                            return [2 /*return*/, false];
                        }
                        _b = this.flags.force;
                        if (_b) return [3 /*break*/, 5];
                        return [4 /*yield*/, helpers_1.default.createTextPrompt('This operation cannot be undone. Confirm deleting selected project by typing its name:', 'Project name must not be empty.')];
                    case 4:
                        _b = (_c.sent());
                        _c.label = 5;
                    case 5:
                        name = _b;
                        if (!this.flags.force && name !== projectName) {
                            this.result.addTextLine('The names do not match.');
                            return [2 /*return*/, false];
                        }
                        this.result.liveTextLine("Unpublishing project " + projectName + "...");
                        query = this.flags['ftp-only'] ? '?ftp=true' : '';
                        this.options.path = "/project/unpublish/" + projectName + query;
                        _c.label = 6;
                    case 6:
                        _c.trys.push([6, 8, , 9]);
                        return [4 /*yield*/, this.http.delete(this.options)];
                    case 7:
                        result = _c.sent();
                        this.result.addAlert(output_color_1.OutputColor.Green, 'Success', result.body);
                        return [2 /*return*/, true];
                    case 8:
                        err_2 = _c.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not delete " + projectName + ": " + err_2.message);
                        return [2 /*return*/, false];
                    case 9: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.kill = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, names, projectName, _a, projectExists, killType, result, err_3;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _b.sent();
                        if (projects.length === 0) {
                            return [2 /*return*/, this.result.addTextLine('You don\'t have any projects yet.')];
                        }
                        names = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = this.flags.name;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', names)];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        projectName = _a;
                        projectExists = projects.some(function (p) { return p.projectName === projectName; });
                        if (!projectExists)
                            return [2 /*return*/, this.result.addTextLine("Project " + projectName + " not found.")];
                        killType = this.flags.remove ? 'rmkill' : 'kill';
                        this.options.path = "/project/" + killType + "/" + projectName;
                        this.result.liveTextLine('Fetching data...');
                        _b.label = 4;
                    case 4:
                        _b.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, this.http.delete(this.options)];
                    case 5:
                        result = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Green, 'Success', result.body);
                        return [3 /*break*/, 7];
                    case 6:
                        err_3 = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not kill " + projectName + ": " + err_3.message);
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.get = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, choices, projectName, _a, project, result, repoUrlWithNicename, query, err_4;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _b.sent();
                        if (projects.length === 0) {
                            return [2 /*return*/, this.result.addTextLine('You don\'t have any projects yet.')];
                        }
                        choices = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = this.flags.name || this.args[0];
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', choices)];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        projectName = _a;
                        project = projects.find(function (p) { return p.projectName === projectName; });
                        if (!project)
                            return [2 /*return*/, this.result.addTextLine("Project " + projectName + " not found.")];
                        _b.label = 4;
                    case 4:
                        _b.trys.push([4, 10, , 11]);
                        if (!(project.repoUrl && !this.flags.ftp)) return [3 /*break*/, 6];
                        repoUrlWithNicename = project.repoUrl.replace(/^https:\/\//, "https://" + project.user.userNicename + "@");
                        return [4 /*yield*/, this.git.clone(repoUrlWithNicename)];
                    case 5:
                        result = _b.sent();
                        return [3 /*break*/, 9];
                    case 6: return [4 /*yield*/, helpers_1.default.eraseDirectories(path_1.default.join(process.cwd(), projectName))];
                    case 7:
                        _b.sent();
                        query = this.flags.force ? '?force=true' : '';
                        this.options.path = "/project/get/" + projectName + query;
                        return [4 /*yield*/, helpers_1.default.downloadFromFTP(this.http, this.options, process.cwd())];
                    case 8:
                        result = _b.sent();
                        _b.label = 9;
                    case 9:
                        this.result.addAlert(output_color_1.OutputColor.Green, 'Success', result);
                        return [3 /*break*/, 11];
                    case 10:
                        err_4 = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not download " + projectName + ": " + err_4.message);
                        return [3 /*break*/, 11];
                    case 11: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.info = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, names, projectName, _a, projectExists, result, startedAt, killedAt, url, isUp, err_5;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _b.sent();
                        if (projects.length === 0) {
                            return [2 /*return*/, this.result.addTextLine('You don\'t have any projects yet.')];
                        }
                        names = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = this.flags.name || this.args[0];
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', names)];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        projectName = _a;
                        projectExists = projects.some(function (p) { return p.projectName === projectName; });
                        if (!projectExists)
                            return [2 /*return*/, this.result.addTextLine("Project " + projectName + " not found.")];
                        this.options.path = "/project/info/" + projectName;
                        this.result.liveTextLine('Fetching data...');
                        _b.label = 4;
                    case 4:
                        _b.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, this.http.get(this.options)];
                    case 5:
                        result = _b.sent();
                        result = JSON.parse(result.body);
                        startedAt = result.startedAt, killedAt = result.killedAt, url = result.url, isUp = result.isUp;
                        this.result.addAlert(output_color_1.OutputColor.Turquoise, 'Status:', isUp ? 'running' : 'dead');
                        this.result.addAlert(output_color_1.OutputColor.Turquoise, isUp ? 'Started at:' : 'Killed at:', isUp ? startedAt : killedAt);
                        if (isUp) {
                            this.result.addAlert(output_color_1.OutputColor.Turquoise, 'App URL:', url);
                        }
                        return [3 /*break*/, 7];
                    case 6:
                        err_5 = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error:', err_5.message);
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.logs = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, names, projectName, _a, projectExists, _b, lines, tail, follow, followQuery, linesQuery, http, request, result, parsedResult, err_6;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _c.sent();
                        if (projects.length === 0) {
                            return [2 /*return*/, this.result.addTextLine('You don\'t have any projects yet.')];
                        }
                        names = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = this.flags.name;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', names)];
                    case 2:
                        _a = (_c.sent());
                        _c.label = 3;
                    case 3:
                        projectName = _a;
                        projectExists = projects.some(function (p) { return p.projectName === projectName; });
                        if (!projectExists)
                            return [2 /*return*/, this.result.addTextLine("Project " + projectName + " not found.")];
                        _b = this.flags, lines = _b.lines, tail = _b.tail, follow = _b.follow;
                        followQuery = follow ? '?follow=true' : '?follow=false';
                        linesQuery = (lines || tail) ? "&lines=" + (lines || tail) : '';
                        this.options.path = "/project/logs/" + projectName + followQuery + linesQuery;
                        this.result.liveTextLine('Fetching data...');
                        if (!follow) return [3 /*break*/, 4];
                        http = new http_wrapper_1.default();
                        request = http.createRawRequest(this.options, function (response) {
                            response.on('data', function (chunk) { return _this.result.liveTextLine(Buffer.from(chunk).toString('utf8')); });
                            response.on('error', function (err) { throw err; });
                        });
                        request.on('error', function (err) { throw err; });
                        request.end();
                        return [3 /*break*/, 7];
                    case 4:
                        _c.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, this.http.get(this.options)];
                    case 5:
                        result = _c.sent();
                        parsedResult = JSON.parse(result.body);
                        this.result.addTextLine(parsedResult.logs);
                        return [3 /*break*/, 7];
                    case 6:
                        err_6 = _c.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not fetch logs for " + projectName + ": " + err_6.message);
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.restart = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, choices, projectName, _a, project, result, err_7;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _b.sent();
                        if (projects.length === 0) {
                            return [2 /*return*/, this.result.addTextLine('You don\'t have any projects yet.')];
                        }
                        choices = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = this.flags.name;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', choices)];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        projectName = _a;
                        project = projects.find(function (p) { return p.projectName === projectName; });
                        if (!project)
                            return [2 /*return*/, this.result.addTextLine("Project " + projectName + " not found.")];
                        this.options.path = "/project/restart/" + projectName;
                        this.result.liveTextLine('Fetching data...');
                        _b.label = 4;
                    case 4:
                        _b.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, this.http.post(this.options)];
                    case 5:
                        result = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Green, 'Success', result.body);
                        return [3 /*break*/, 7];
                    case 6:
                        err_7 = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not restart project " + projectName + ": " + err_7.message);
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.run = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projects, choices, projectName, _a, project, meta, technology, result, message, err_8;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, this.getBackendProjects()];
                    case 1:
                        projects = _b.sent();
                        if (projects.length === 0) {
                            return [2 /*return*/, this.result.addTextLine('You don\'t have any projects yet.')];
                        }
                        choices = projects.map(function (p) { return ({ name: p.projectName }); });
                        _a = this.flags.name;
                        if (_a) return [3 /*break*/, 3];
                        return [4 /*yield*/, helpers_1.default.createListPrompt('Choose project', choices)];
                    case 2:
                        _a = (_b.sent());
                        _b.label = 3;
                    case 3:
                        projectName = _a;
                        project = projects.find(function (p) { return p.projectName === projectName; });
                        if (!project)
                            return [2 /*return*/, this.result.addTextLine("Project " + projectName + " not found.")];
                        meta = project.projectMeta.find(function (m) { return m.metaKey === '_backend_technology'; });
                        technology = meta ? meta.metaValue : undefined;
                        this.options.path = "/project/run/" + technology + "/" + projectName;
                        this.result.liveTextLine('Fetching data...');
                        _b.label = 4;
                    case 4:
                        _b.trys.push([4, 6, , 7]);
                        return [4 /*yield*/, this.http.post(this.options)];
                    case 5:
                        result = _b.sent();
                        message = JSON.parse(result.body).message;
                        this.result.addTextLine(message);
                        return [3 /*break*/, 7];
                    case 6:
                        err_8 = _b.sent();
                        this.result.addAlert(output_color_1.OutputColor.Red, 'Error', "Could not run project " + projectName + ": " + err_8.message);
                        return [3 /*break*/, 7];
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    BackendReceiver.prototype.askForProjectName = function () {
        return __awaiter(this, void 0, void 0, function () {
            var projectName;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, helpers_1.default.createTextPrompt('Enter project name', 'Project name must not be empty.')];
                    case 1:
                        projectName = _a.sent();
                        this.context.mdbConfig.setValue('projectName', projectName);
                        this.context.mdbConfig.save();
                        return [2 /*return*/];
                }
            });
        });
    };
    return BackendReceiver;
}(receiver_1.default));
exports.default = BackendReceiver;
